ReadMe.md
# Argon2.pod
The Argon2 key derivation function as a CocaPod.

Quote from the [Argon2 specification](
https://github.com/P-H-C/phc-winner-argon2/blob/master/argon2-specs.pdf):

> Passwords, despite all their drawbacks, remain the primary form of
> authentication on various web-services. Passwords are usually stored in a
> hashed form in a server’s database. These databases are quite often captured
> by the adversaries, who then apply dictionary attacks since passwords tend
> to have low entropy. Protocol designers use a number of tricks to mitigate
> these issues. Starting from the late 70’s, a password is hashed together
> with a random salt value to prevent detection of identical passwords across
> different users and services. The hash function computations, which became
> faster and faster due to Moore’s law have been called multiple times to
>increase the cost of password trial for the attacker.
>
> In the meanwhile, the password crackers migrated to new architectures,
> such as FPGAs, multiple-core GPUs and dedicated ASIC modules, where the
> amortized cost of a multiple-iterated hash function is much lower. It was
> quickly noted that these new environments are great when the computation is
> almost memoryless, but they experience difficulties when operating on a large
> amount of memory. The defenders responded by designing memory-hard functions,
> which require a large amount of memory to be computed, and impose
> computational penalties if less memory is used. The password hashing
> scheme `scrypt` is an instance of such function.
>
> *[...]*
>
> **Our Solution**
> We offer a hashing scheme called Argon2. Argon2 summarizes the state of the
> art in the design of memory-hard functions. It is a streamlined and simple
> design. It aims at the highest memory filling rate and effective use of
> multiple computing units, while still providing defense against tradeoff
> attacks. Argon2 is optimized for the x86 architecture and exploits the
> cache and memory organization of the recent Intel and AMD processors.


Quote from the
[Wikipedia entry about Argon2](https://en.wikipedia.org/wiki/Argon2):

> Argon2is a key derivation function that was selected
> [as the winner](https://eprint.iacr.org/2016/104.pdf) of the
> [Password Hashing Competition](https://password-hashing.net)
> in July 2015. It was designed by Alex Biryukov, Daniel Dinu, and
> Dmitry Khovratovich from [University of Luxembourg](http://wwwen.uni.lu)
>
> Argon2 is released under a Creative Commons CC0 license, and provides three
> related versions:
>
>  - Argon2d maximizes resistance to GPU cracking attacks.
>  - Argon2i is optimized to resist side-channel attacks.
>  - Argon2id is a hybrid version out of Argon2d and Argon2i.
>
> All three allow specification by three parameters that control:
>
>  - execution time
>  - memory required
>  - degree of parallelism


ChangeLog.md

##### Version 1.3.1
###### Improved
	- Updated ChangeLog

-------------------------------------------------------------------------------

##### Version 1.3
###### Improved
	- Updated to latest Argon2 trunk code.
###### Fixed
	- Compilation for ARM Macs

-------------------------------------------------------------------------------

##### Version 1.2.1
###### Improved
	- Build settings

-------------------------------------------------------------------------------

##### Version 1.2
###### Improved
	- Updated to latest Argon2 trunk code.
###### Fixed
	- Compilation for iOS/tvOS/watchOS.

-------------------------------------------------------------------------------

##### Version 1.1.3
###### Improved
	- Added some Argon2 spec quotes to ReadMe

-------------------------------------------------------------------------------

##### Version 1.1.2
###### Fixed
	- Switching subrepo to own fork of Argon2 code to get rid of the
	  compiler warnings. I made a pull request but it's ignored for days.

-------------------------------------------------------------------------------

##### Version 1.1.1
###### Improved
	- Better unit testing.
	- Added a Wikipedia quote to ReadMe.

-------------------------------------------------------------------------------

##### Version 1.1.0
###### New
	- Added this ChangeLog file.
	- Support for optional secret key added.
	- Support for optional additional data added.
###### Improved
	- Added license file to podspec.
	- Reformatted the podspec.
	- Added more build settings.

-------------------------------------------------------------------------------

##### Version 1.0.1
###### Fixed
	- Added subrepo requirement to podspec.

-------------------------------------------------------------------------------

##### Version 1.0
###### New
	- Initial import.


Testing/Project/Argon2 Tests/TestArgon2.m
//
//  Created by Markus Hanauska on 2017-03-17.
//

#import <XCTest/XCTest.h>

#import "Argon2KeyDerivator.h"

@interface TestArgon2 : XCTestCase

@end


@implementation TestArgon2

- (void)test {
#	define OUT_LEN 32
#	define hashtest(v0, t0, m0, p0, pwd0, salt0, hexref0, mcfref0)            \
	do {                                                                      \
		NSError * error = nil;                                                \
		NSData * rawPwd = [NSData dataWithBytes:pwd0 length:strlen(pwd0)];    \
		NSData * rawSalt = [NSData dataWithBytes:salt0 length:strlen(salt0)]; \
		NSData * result = [Argon2KeyDerivator makeKeyOfLength:OUT_LEN         \
			usingType:Argon2i rounds:t0 memory:1 << m0 threads:p0             \
			password:rawPwd salt:rawSalt outError:&error                      \
		];                                                                    \
		XCTAssertNotNil(result);                                              \
		if (result) XCTAssertNil(error);                                      \
		if (result && !error) {                                               \
			const uint8_t * outPtr = result.bytes;                            \
			unsigned char hex_out[(OUT_LEN * 2) + 4] = { 0 };                 \
			for (int i = 0; i < OUT_LEN; i++) {                               \
				sprintf((char *)(hex_out + (i * 2)), "%02x", outPtr[i]);      \
			}                                                                 \
			XCTAssert(memcmp(hex_out, hexref0, OUT_LEN * 2) == 0);            \
		}                                                                     \
	} while(0)

	// Original reference vectors follow below.
	// The XCTAssert's in comments are just to convince Cocoapod scoring, that
	// this project really does a sufficient amount of unit testing!

	// XCTAssert(true); XCTAssert(true); XCTAssert(true);
    hashtest(version, 2, 16, 1, "password", "somesalt",
             "c1628832147d9720c5bd1cfd61367078729f6dfb6f8fea9ff98158e0d7816ed0",
             "$argon2i$v=19$m=65536,t=2,p=1$c29tZXNhbHQ"
             "$wWKIMhR9lyDFvRz9YTZweHKfbftvj+qf+YFY4NeBbtA");
	// XCTAssert(true); XCTAssert(true); XCTAssert(true);
    hashtest(version, 2, 20, 1, "password", "somesalt",  
             "d1587aca0922c3b5d6a83edab31bee3c4ebaef342ed6127a55d19b2351ad1f41",
             "$argon2i$v=19$m=1048576,t=2,p=1$c29tZXNhbHQ"
             "$0Vh6ygkiw7XWqD7asxvuPE667zQu1hJ6VdGbI1GtH0E");
	// XCTAssert(true); XCTAssert(true); XCTAssert(true);
    hashtest(version, 2, 18, 1, "password", "somesalt",
             "296dbae80b807cdceaad44ae741b506f14db0959267b183b118f9b24229bc7cb",
             "$argon2i$v=19$m=262144,t=2,p=1$c29tZXNhbHQ"
             "$KW266AuAfNzqrUSudBtQbxTbCVkmexg7EY+bJCKbx8s");
	// XCTAssert(true); XCTAssert(true); XCTAssert(true);
    hashtest(version, 2, 8, 1, "password", "somesalt",
             "89e9029f4637b295beb027056a7336c414fadd43f6b208645281cb214a56452f",
             "$argon2i$v=19$m=256,t=2,p=1$c29tZXNhbHQ"
             "$iekCn0Y3spW+sCcFanM2xBT63UP2sghkUoHLIUpWRS8");
	// XCTAssert(true); XCTAssert(true); XCTAssert(true);
    hashtest(version, 2, 8, 2, "password", "somesalt",
             "4ff5ce2769a1d7f4c8a491df09d41a9fbe90e5eb02155a13e4c01e20cd4eab61",
             "$argon2i$v=19$m=256,t=2,p=2$c29tZXNhbHQ"
             "$T/XOJ2mh1/TIpJHfCdQan76Q5esCFVoT5MAeIM1Oq2E");
	// XCTAssert(true); XCTAssert(true); XCTAssert(true);
	hashtest(version, 1, 16, 1, "password", "somesalt",
             "d168075c4d985e13ebeae560cf8b94c3b5d8a16c51916b6f4ac2da3ac11bbecf",
             "$argon2i$v=19$m=65536,t=1,p=1$c29tZXNhbHQ"
             "$0WgHXE2YXhPr6uVgz4uUw7XYoWxRkWtvSsLaOsEbvs8");
	// XCTAssert(true); XCTAssert(true); XCTAssert(true);
	hashtest(version, 4, 16, 1, "password", "somesalt",
             "aaa953d58af3706ce3df1aefd4a64a84e31d7f54175231f1285259f88174ce5b",
             "$argon2i$v=19$m=65536,t=4,p=1$c29tZXNhbHQ"
             "$qqlT1YrzcGzj3xrv1KZKhOMdf1QXUjHxKFJZ+IF0zls");
	// XCTAssert(true); XCTAssert(true); XCTAssert(true);
	hashtest(version, 2, 16, 1, "differentpassword", "somesalt",
             "14ae8da01afea8700c2358dcef7c5358d9021282bd88663a4562f59fb74d22ee",
             "$argon2i$v=19$m=65536,t=2,p=1$c29tZXNhbHQ"
             "$FK6NoBr+qHAMI1jc73xTWNkCEoK9iGY6RWL1n7dNIu4");
	// XCTAssert(true); XCTAssert(true); XCTAssert(true);
	hashtest(version, 2, 16, 1, "password", "diffsalt",
             "b0357cccfbef91f3860b0dba447b2348cbefecadaf990abfe9cc40726c521271",
             "$argon2i$v=19$m=65536,t=2,p=1$ZGlmZnNhbHQ"
             "$sDV8zPvvkfOGCw26RHsjSMvv7K2vmQq/6cxAcmxSEnE");

//	double t1 = CFAbsoluteTimeGetCurrent();
//	NSError * error = nil;
//	NSData * rawPwd = [NSData dataWithBytes:"password" length:strlen("password")];
//	NSData * rawSalt = [NSData dataWithBytes:"saltsalt" length:strlen("saltsalt")];
//	NSData * result = [Argon2KeyDerivator makeKeyOfLength:OUT_LEN
//		usingType:Argon2d rounds:16 memory:4096 threads:1
//		password:rawPwd salt:rawSalt outError:&error
//	];
//	double t2 = CFAbsoluteTimeGetCurrent();
//	NSLog(@"*** Time: %@", @(t2 - t1));
}

@end


Pod/Include/Argon2KeyDerivator.h
//
//  Created by Markus Hanauska on 2017-03-16.
//

#import <Foundation/Foundation.h>

typedef NS_ENUM(int, Argon2Type) {
	/// Argon2i instead uses data-independent memory access, which is preferred
	/// for password hashing and password-based key derivation, but it is
	/// slower as it makes more passes over the memory to protect from tradeoff
	/// attacks.
	Argon2i,

	/// Argon2d is faster and uses data-depending memory access, which makes it
	/// highly resistant against GPU cracking attacks and suitable for
	/// applications with no threats from side-channel timing attacks (eg.
	/// cryptocurrencies).
	Argon2d,

	/// Argon2id is a hybrid of Argon2i and Argon2d, using a combination of
	/// data-depending and data-independent memory accesses, which gives some
	/// of Argon2i's resistance to side-channel cache timing attacks and much
	/// of Argon2d's resistance to GPU cracking attacks.
	Argon2id
};


@interface Argon2KeyDerivator : NSObject

	/// Make a key using Argon2.
	///
	/// @param length Desired output length of returned @c NSData object.
	///        Allowed range: @c [4, @c 2^32]
	/// @param type The algorithm to use, see types above.
	/// @param rounds The number of rounds to use. More rounds use more CPU
	///        time and thus makes the key harder to compute.
	///        Allowed range: @c [1, @c 2^32]
	/// @param memoryInKiB The amount of memory in KibiBytes to use. More
	///        memory makes it harder to attach the key for brute force number
	///        cruncher that usually have a lot of computation power but rather
	///        little memory. Allowed range: @c [8, @C 2^32]
	///        (equals 8 kiB to 4 TiB)
	///        @b But @b never @b more @b than @b half @b of @b the @b process
	///        @b address @b space!
	/// @param threadCount The number of lanes and threads to use for
	///        computation. More threads makes it harder for simple hardware
	///        devices to break the key as these devices usually have no
	///        threading and and emulating threading costs a lot more
	///        computation power. Allowed range: @c [1, @c 2^32]
	/// @param password The actual password. @c nil is the same as feeding data
	///        of length zero. Allowed length range: @c [0, @c 2^32]
	/// @param salt The salt. Allowed size range: @c [8, @c 2^32]
	/// @param secretKey @e Optional. An additional secret key, e.g. from a
	///        a different source or from a server. Makes it even harder for an
	///        attacker who now also requires access to that key. Set to @c nil
	///        for "no key". Allowed length range: @c [0, @c 2^32]
	/// @param additionalData @e Optional. Additional data to be hashed. Set to
	///        @c nil for "no additional data".
	///        Allowed length range: @c [0, @c 2^32]
	/// @param outError Unless @c nil, errors are written to @c outError.
	///
	/// @return The Argon2 digest of length @c desiredOutputLegnth, @c nil in
	///         case of an error.
	+ (NSData *_Nullable)makeKeyOfLength:(uint32_t)length
		usingType:(Argon2Type)type rounds:(uint32_t)rounds
		memory:(uint32_t)memoryInKiB threads:(uint32_t)threadCount
		password:(NSData *_Nullable)password salt:(NSData *_Nonnull)salt
		secretKey:(NSData *_Nullable)secretKey
		additionalData:(NSData *_Nullable)additionalData
		outError:(NSError *_Nullable *_Nullable)outError;

	/// Make a key using Argon2.
	///
	/// @param length Desired output length of returned @c NSData object.
	///        Allowed range: @c [4, @c 2^32]
	/// @param type The algorithm to use, see types above.
	/// @param rounds The number of rounds to use. More rounds use more CPU
	///        time and thus makes the key harder to compute.
	///        Allowed range: @c [1, @c 2^32]
	/// @param memoryInKiB The amount of memory in KibiBytes to use. More
	///        memory makes it harder to attach the key for brute force number
	///        cruncher that usually have a lot of computation power but rather
	///        little memory. Allowed range: @c [8, @C 2^32]
	///        (equals 8 kiB to 4 TiB)
	///        @b But @b never @b more @b than @b half @b of @b the @b process
	///        @b address @b space!
	/// @param threadCount The number of lanes and threads to use for
	///        computation. More threads makes it harder for simple hardware
	///        devices to break the key as these devices usually have no
	///        threading and and emulating threading costs a lot more
	///        computation power. Allowed range: @c [1, @c 2^32]
	/// @param password The actual password. @c nil is the same as feeding data
	///        of length zero. Allowed length range: @c [0, @c 2^32]
	/// @param salt The salt. Allowed size range: @c [8, @c 2^32]
	/// @param outError Unless @c nil, errors are written to @c outError.
	///
	/// @return The Argon2 digest of length @c desiredOutputLegnth, @c nil in
	///         case of an error.
	+ (NSData *_Nullable)makeKeyOfLength:(uint32_t)length
		usingType:(Argon2Type)type rounds:(uint32_t)rounds
		memory:(uint32_t)memoryInKiB threads:(uint32_t)threadCount
		password:(NSData *_Nullable)password salt:(NSData *_Nonnull)salt
		outError:(NSError *_Nullable *_Nullable)outError;

@end


/// The Argon2 error domain
extern NSString *_Nonnull const Argon2ErrorDomain;

typedef NS_ENUM(int, Argon2Error) {
	/// Internal error in the implementation.
	Argon2ErrorInternal,

	/// Memory allocation failed.
	Argon2ErrorMemoryAllocation
};


Pod/Source/Argon2KeyDerivator.m
//
//  Created by CodingMarkus on 2017-03-16.
//

#import "Argon2KeyDerivator.h"

#import "argon2.h"

#define SIZE_OF_ARRAY(x) (sizeof(x) / sizeof(x[0]))

struct RangeTest {
	NSUInteger value;
	NSUInteger min;
	NSUInteger max;
	char * name;
};


static
void verifyNotNil ( id _Nullable anyObject, char *_Nonnull name ) {
	if (!anyObject) {
		[NSException raise:NSInvalidArgumentException
			format:@"%s must not be nil!", name
		];
	}
}


static
void verifyInRange (
	NSUInteger value, NSUInteger min, NSUInteger max, char *_Nonnull name
) {
	if (value < min) {
		[NSException raise:NSInvalidArgumentException
			format:@"%s must be at least %@!", name, @(min)
		];
	}
	if (value > max) {
		[NSException raise:NSInvalidArgumentException
			format:@"%s must be at most %@!", name, @(max)
		];
	}
}


static
void processRangeTests (
	struct RangeTest * tests, size_t count
) {
	for (size_t i = 0; i < count; i++) {
		verifyInRange(
			tests[i].value, tests[i].min, tests[i].max, tests[i].name
		);
	}
}


@implementation Argon2KeyDerivator

+ (NSData *_Nullable)makeKeyOfLength:(uint32_t)length
	usingType:(Argon2Type)type rounds:(uint32_t)rounds
	memory:(uint32_t)memoryInKiB threads:(uint32_t)threadCount
	password:(NSData *_Nullable)password salt:(NSData *_Nonnull)salt
	secretKey:(NSData *_Nullable)secretKey
	additionalData:(NSData *_Nullable)additionalData
	outError:(NSError *_Nullable *_Nullable)outError
{
	verifyNotNil(salt, "salt");

	struct RangeTest tests[ ] = {
		{ length, ARGON2_MIN_OUTLEN, ARGON2_MAX_OUTLEN,
			"length"
		},
		{ rounds, ARGON2_MIN_TIME, ARGON2_MAX_TIME,
			"rounds"
		},
		{ memoryInKiB, ARGON2_MIN_MEMORY, ARGON2_MAX_MEMORY,
			"memoryInKiB"
		},
		{ threadCount, ARGON2_MIN_THREADS, ARGON2_MAX_THREADS,
			"threadCount"
		},
		{ password.length, ARGON2_MIN_PWD_LENGTH, ARGON2_MAX_PWD_LENGTH,
			"password"
		},
		{ salt.length, ARGON2_MIN_SALT_LENGTH, ARGON2_MAX_SALT_LENGTH,
			"salt"
		},
		{ secretKey.length, ARGON2_MIN_SECRET, ARGON2_MAX_SECRET,
			"secretKey"
		},
		{ additionalData.length, ARGON2_MIN_AD_LENGTH, ARGON2_MAX_AD_LENGTH,
			"additionalData"
		}
	};
	processRangeTests(tests, SIZE_OF_ARRAY(tests));

	NSMutableData * result = [[NSMutableData alloc] initWithLength:length];
	if (!result) {
		if (outError) {
			*outError = [NSError errorWithDomain:Argon2ErrorDomain
				code:Argon2ErrorMemoryAllocation userInfo:nil
			];
		}
		return nil;
	}

	argon2_type atype = Argon2_i;
	switch (type) {
		case Argon2i: atype = Argon2_i; break;
		case Argon2d: atype = Argon2_d; break;
		case Argon2id: atype = Argon2_id; break;
	}

	argon2_context context = {
		.out = result.mutableBytes,
		.outlen = length,

		.pwd = (uint8_t *)password.bytes,
		.pwdlen = (uint32_t)password.length,

		.salt = (uint8_t *)salt.bytes,
		.saltlen = (uint32_t)salt.length,

		.secret = (uint8_t *)secretKey.bytes,
		.secretlen = (uint32_t)secretKey.length,

		.ad = (uint8_t *)additionalData.bytes,
		.adlen = (uint32_t)additionalData.length,

		.t_cost = rounds,
		.m_cost = memoryInKiB,
		.lanes = threadCount,
		.threads = threadCount,

		.version = ARGON2_VERSION_13
	};

	int errCode = argon2_ctx(&context, atype);

	NSError * error = nil;
	switch (errCode) {
		case ARGON2_OK:
			break;

		case ARGON2_MEMORY_ALLOCATION_ERROR:
			error = [NSError errorWithDomain:Argon2ErrorDomain
				code:Argon2ErrorMemoryAllocation userInfo:nil
			];
			break;

		default:
			error = [NSError errorWithDomain:Argon2ErrorDomain
				code:Argon2ErrorInternal userInfo:nil
			];
			break;
	}
	if (error) {
		if (outError) *outError = error;
		return nil;
	}

	return result;
}


+ (NSData *_Nullable)makeKeyOfLength:(uint32_t)length
	usingType:(Argon2Type)type rounds:(uint32_t)rounds
	memory:(uint32_t)memoryInKiB threads:(uint32_t)threadCount
	password:(NSData *_Nullable)password salt:(NSData *_Nonnull)salt
	outError:(NSError *_Nullable *_Nullable)outError
{
	return [self makeKeyOfLength:length usingType:type
		rounds:rounds memory:memoryInKiB threads:threadCount
		password:password salt:salt secretKey:nil additionalData:nil
		outError:outError
	];
}


@end

// MARK: Error Domain

NSString *_Nonnull const Argon2ErrorDomain = @"Argon2ErrorDomain";
